sympy/functions/elementary/hyperbolic.py

```diff
@@ -269,9 +269,7 @@ def _eval_is_finite(self):
         return arg.is_finite
 
     def _eval_is_zero(self):
-        arg = self.args[0]
-        if arg.is_zero:
-            return True
+        return (self.args[0]/(S.Pi*I)).is_integer
```

Comment:
This kind of thing is very tricky and should be avoided if possible. What this does is create a new Basic instance as part of an assumptions query. There are several reasons this can be problematic:

1. The assumptions system should lie on a level below evaluation so that it can be called as part of evaluation. Invoking evaluation as part of the assumptions system itself is circular.
2. Apart from making a messy design the circularity creates a risk of infinite recursion where an object directly or indirectly recreates itself as part of the assumptions query.
3. The `is_zero` predicate is called in many places e.g. `Mul.flatten` and as a result any change can have wide performance impact.
4. If the expression is large then putting it into a Mul can be expensive as `Mul.flatten` needs to rerun.

Also it is incorrect to conclude that `(self.args[0]/(S.Pi*I)).is_integer = False` implies `self.is_zero = False` because of e.g. `zoo`, `nan` etc. There needs to be a guard here like checking `self.args[0].is_complex`.

Reply:
Regarding incorrectness, see my other comment. Can really `sin(zoo)` be considered to possibly be zero?

As far as I can tell, there is no need to check for complex here as a real value will return an imaginary value that is never integer.

For the rest of the comments, I see the potential risk of circular execution. I just wonder how to get around that `cosh(5*I*S.Pi/2, evaluate=False).is_zero` currently returns `False`.

Is there a "safe" way to extract `Pi` from the expression?

Reply:
There's a function somewhere called something like `peel_off_pi`.

Reply:
It seems that the tests run by `_peeloff_pi` are almost the same: https://github.com/sympy/sympy/blob/281b1904acc9e15ed04402cff33ebd6ef7812026/sympy/functions/elementary/trigonometric.py#L127-L129

Reply:
That can definitely be improved. In the meantime it would be better if all places doing similar things use the same function so that it can be improved in one place.